Display icons 229

adrientremblay opened the pull request:
# Assignee Section

## Description
Revamp the battery voltage and temperature indicators on the navbar of the GUI.  Also adds modal popups, colour changes, and a sound when voltage or temp is out of acceptable range.  Also upgrades from fontawsome 4 to 5.

closes #229 

Approval from at least one software team lead is necessary before merging.

# Reviewer Section

Aside from local testing and the General Integration Test it is implied that static analysis should be included in the verification process.

- [x] Local Test Performed Successfully
- [ ] [General Integration Test](https://docs.google.com/document/d/1ug0CpA1cIzURP8DDFSvCt2CEJJSwJ6Ta6B1LG_hYk6I/edit) Performed Successfully

For Pull Requests that do not include code changes, it is not required to perform the tests above.

# Local Testing Guide

## Voltage Indicator Test
tests can be run using
`rostopic pub battery_voltage std_msgs/Float32 <voltage>`

1. go into acceptable range 15V (indicator should be white)
2. go into warning zone 16V (indicator should be orange)
3. go into unacceptable range 17V (indicator should be red)
4. go into acceptable range within leeway 16.5 (this is in the acceptable range but the indicator should remain red)
5. go into warning zone 16V (indicator should be orange)
6. go into acceptable range 15V (indicator should be white)
7.  go into warning zone 13V (indicator should be orange)
8. go into unacceptable zone 12V (indicator should be red)
9. go into acceptable range within leeway 12.3 (this is in the acceptable range but the indicator should remain red)
10. go into acceptable range 15V (indicator should be white)


## Temperature Indicators Test
tests can be run using
`rostopic pub /battery_temps geometry_msgs/Point <T1> <T2> <T3>`

1. go into acceptable range 70 °C (indicator should be white)
2. go into warning zone 80 °C (indicator should be orange)
3. go into unacceptable range 90 °C (indicator should be red)
4. go into acceptable range within leeway 84.5 °C (this is in the acceptable range but the indicator should remain red)
5. go into warning zone 80 °C (indicator should be orange)
6. go into acceptable range 70 °C (indicator should be white)
7.  go into warning zone 5 °C (indicator should be orange)
8. go into unacceptable zone -5 °C (indicator should be red)
9. go into acceptable range within leeway 0.5 °C (this is in the acceptable range but the indicator should remain red)
10. go into acceptable range 70 °C (indicator should be white)

MartensCedric commented on robot/basestation/static/js/roslibjs/ros-helpers.js:
```diff
@@ -535,3 +578,19 @@ function getRoverIP (callback) {
+  $("#navModalTitle").text(title)
+  $("#navModalBody").text(body)
+  $("#navModal").modal({show: true})
+}
```
This belongs probably in `navbar.js`

MartensCedric commented on robot/basestation/static/js/roslibjs/ros-helpers.js:
```diff
@@ -167,9 +187,32 @@ function initRosWeb () {
+        }
+      }
+
+      $('#battery-temp-' + i).css({'color': color})
```
Ah I see why you used a for loop now. Then you should not iterate the values but the jquery objects so you won't even have to do that

MartensCedric commented:
@adrientremblay know that ROS publishes often to that /battery_voltage thing. Imagine if the battery heats A LOT. What do we want to do? We want to turn the rover OFF. Now imagine that since the rover is "spamming" these `/battery_voltage` values. Your code is constantly checking if it should play a sound or not.

Now if the battery does get indeed very hot (and stays hot) then the GUI will be spamming sounds non-stop and you won't be able to close the modal dialog without having a new one pop in your face lol. This is kind of a design issue though. I need thoughts @PeterGhimself

adrientremblay commented on robot/basestation/static/js/roslibjs/ros-helpers.js:
```diff
@@ -167,9 +187,32 @@ function initRosWeb () {
+          navModalMessage('Warning: Battery temperature too high', 'Decreace temperature')
+        } else if (temp < MIN_TEMP) {
+          navModalMessage('Warning: Battery temperature too low', 'Increace temperature')
+        }
```
did u want me to use an image or is it fine with a keep fontawesome? :/

MartensCedric replied:
@adrientremblay  You should keep the threads to one topic only because once the issue is resolve we can close the thread with GitHub.

But I don't mind if you keep it to fontawesome as long as whatever what was used before with 4.7 isnt affected (since you removed it). But it seems to be fine

adrientremblay commented:
@MartensCedric ok sure.

adrientremblay commented:
@MartensCedric @PeterGhimself I made the changes you suggested.  I removed the fontawesome files for regular and brands which weren't being used, moved my js functions to where they should be, and made a template for the modal.  I also found a simple solution to the issue where the fluctuating voltage/temperature would cause a million modal popups and sounds when its near a critical value (max or min).  

What I did is make it so the modal will only activate when a label goes from white(acceptable value) to red(unacceptable value).  Once red the value will have to go back to the critical value and then plus some.  I added VOLTAGE_LEEWAY and TEMP_LEEWAY to represent this buffer zone size.  

So for example say a temperature goes above 85 to say 87 -> it's label becomes red and a modal pops up.  Then it goes back down to 85 -> nothing happens.  it goes down to 84.5 nothing happens(as TEMP_LEEWAY is 1).  Only when the values goes below 84 will the label be reset to white indicating an acceptable temperature.

Let me know guys if this solution makes sense.  :)

MartensCedric commented:
Sounds Great! I'll look at it more in depth on Saturday with you! Congrats
on the good progress!

On Thu, Jan 9, 2020, 11:47 PM Adrien Tremblay, <notifications@github.com>
wrote:

> @MartensCedric <https://github.com/MartensCedric> @PeterGhimself
> <https://github.com/PeterGhimself> I made the changes you suggested. I
> removed the fontawesome files for regular and brands which weren't being
> used, moved my js functions to where they should be, and made a template
> for the modal. I also found a simple solution to the issue where the
> fluctuating voltage/temperature would cause a million modal popups and
> sounds when its near a critical value (max or min).
>
> What I did is make it so the modal will only activate when a label goes
> from white(acceptable value) to red(unacceptable value). Once red the value
> will have to go back to the critical value and then plus some. I added
> VOLTAGE_LEEWAY and TEMP_LEEWAY to represent this buffer zone size.
>
> So for example say a temperature goes above 85 to say 87 -> it's label
> becomes red and a modal pops up. Then it goes back down to 85 -> nothing
> happens. it goes down to 84.5 nothing happens(as TEMP_LEEWAY is 1). Only
> when the values goes below 84 will the label be reset to white indicating
> an acceptable temperature.
>
> Let me know guys if this solution makes sense. :)
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/space-concordia-robotics/robotics-prototype/pull/272?email_source=notifications&email_token=AEOB46LV6Q36IUSF7IKZM73Q474V3A5CNFSM4KEQVHL2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEISVIAY#issuecomment-572871683>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEOB46ODMCFPNIVNOEJK7FLQ474V3ANCNFSM4KEQVHLQ>
> .
>

adrientremblay commented:
<img width="538" alt="example" src="https://user-images.githubusercontent.com/8109538/72209769-0efad800-3480-11ea-82f2-93490ea8e33e.png">

This diagram is an example of how my algorithm works in the example i described earlier.  Green represents the zone where the value will be turn from unacceptable to acceptable and the red is where the value will be marked as unacceptable.  Notice there is a blank zone in between these two values where the state of the value will not change.

PeterGhimself commented on robot/basestation/static/js/roslibjs/ros-helpers.js:
```diff
@@ -2,6 +2,15 @@ REQUEST_TIMEOUT = 3000
 
+// minimum and maximum acceptable battery voltages (volts)
+MIN_VOLTAGE = 12.5
+MAX_VOLTAGE = 16.9
```
Please change this to 16.8 and make sure that the error modal only appears when it is strictly _greater_ than 16.8 (so 16.8 is allowed). The rest are fine, it's just this one that needs to be changed in this sense.

adrientremblay commented:
Need to add tests and the disable var

PeterGhimself commented on tests/unit/branch_name_test.py:
```diff
@@ -0,0 +1,71 @@
+import branch_name_verification_hook as branch_name_verification_hook
```
this is already in master, so why is it being added in this PR? it has nothing to do with #229

adrientremblay replied:
I have no idea why I have these files

adrientremblay replied:
@PeterGhimself is branch_name_test.py the only file i should delete.  There are a bunch of other files that I didn't write that are in my PR for whatever reason such as import_test.py, branch_name_verification_hook.py, pytest.ini  ... Should I delete them all?

MartensCedric replied:
`git diff f357b` to show where the issue happens. Im very confused to what happened and why GH says its a conflict

adrientremblay replied:
@PeterGhimself @MartensCedric Cedric readded the module (thumbs up emoji)

PeterGhimself commented on robot/basestation/static/js/roslibjs/ros-helpers.js:
```diff
@@ -2,6 +2,17 @@ REQUEST_TIMEOUT = 3000
+MAX_TEMP = 85
+TEMP_LEEWAY = 1
+// variable to toggle unacceptable voltage and temperature indicators
+ALERT_ENABLE = false;
```
shouldn't this be true by default? also CS --> no semicolons

adrientremblay replied:
im small brain

PeterGhimself commented:
By the way we should probably create a new issue since it's out of scope for this PR, but to add custom controls on these ranges for min/max as well as the feature toggle.
It's good to have the option in the code already, but when I think about the operations at comp it would make the most sense that if we _do_ need to disable the feature for whatever reason or adjust values to compensate for something it would be much more convenient to manage that through the UI rather than finding that particular line of code, changing it (without typo) and `ctrl + shift + R`-ing the web page to see the change take place.

adrientremblay commented on robot/basestation/static/js/roslibjs/ros-helpers.js:
```diff
@@ -158,7 +169,23 @@ function initRosWeb () {
   })
   battery_voltage_listener.subscribe(function (message) {
-    $('#battery-voltage').text(message.data.toFixed(2))
+    let voltage = message.data.toFixed(2)
```
yepp!

adrientremblay commented on robot/basestation/static/js/app.js:
```diff
@@ -252,3 +252,10 @@ function pingOdroid (timeoutVal = REQUEST_TIMEOUT) {
+plays an error sound once -> warning will not play if the document has not been interacted with yet
+*/
+function errorSound (){
+  new Audio('../../static/sound/error.mp3').play()
```
Not sure what this means :/ sorry

adrientremblay replied:
@PeterGhimself

PeterGhimself replied:
[google is friend](https://en.wikipedia.org/wiki/Magic_number_(programming))

adrientremblay replied:
i created a variable to store the path.  Hope thats what you meant :-<

MartensCedric replied:
Yeah that's what he meant

adrientremblay commented on robot/basestation/static/js/roslibjs/ros-helpers.js:
```diff
@@ -158,7 +169,23 @@ function initRosWeb () {
+      }
+    } else if ($('#battery-voltage').attr('acceptable') === '0' && voltage < MAX_VOLTAGE - VOLTAGE_LEEWAY && voltage > MIN_VOLTAGE + VOLTAGE_LEEWAY) {
+      $('#battery-voltage').attr('acceptable', '1')
+      $('#battery-voltage').css({'color': 'white'})
```
so that function doesnt work cause it changes the 'background-color' property instead of the 'color' property.  I'll just create my own function in helpers.js

MartensCedric commented:
LGTM

MartensCedric commented:
@PeterGhimself can this be approved?

vashmata commented:
Why does it say to disconnect the BMS when there's a voltage error? Shouldn't it always be plugged into the battery @TheRealJeff 
![image](https://user-images.githubusercontent.com/32688392/74095393-c570cd80-4abd-11ea-95f5-1f528c9fd057.png)

On my laptop I'm not getting any audio when the popup appears.

There's an error for the battery temp being too hot but not for it being too cold. Not sure what the limits are as I'm not reading the code.

It would also be nice to specify which temp sensor(s) is(are) out of safe range

adrientremblay commented:
@vashmata I checked already with @TheRealJeff a while ago for the modal messages.  So they should be good.  I know nothing about the hardware of the rover so maybe some confirmation about messages would be good.  

Theres no audio because you need to interact with the site before chrome is allowed to play sounds.  So interact with something then trigger the popup and the sound should play.  

The min temperature is 0 degrees.  It'll popup if you give a negative temp.

I'll make the changes so the modal displays  which temperature sensor is out of range.  (T1 to T3)

vashmata commented:
Thanks for the update. I'll check with jeff about the bms
 
> Theres no audio because you need to interact with the site before chrome is allowed to play sounds. So interact with something then trigger the popup and the sound should play.

I was with Cedric when using the GUI and he clicked on things a bunch and then I published a new invalid voltage and it made a popup with no audio. Maybe it's cause I switched out of chrome to publish the message?
 
> The min temperature is 0 degrees. It'll popup if you give a negative temp.

I sent it negative temperatures

adrientremblay commented:
OK perfect.  The sound not playing is pretty weird.  If the sound is being suppressed because you haven't interacted with the page then there should be a message in console.  I'd want to check console if that message pops up or not, if it doesn't i guess its something else.  I'll look into the negative temperature thing and double check I can send negative values.

adrientremblay commented:
I just tested and It works on my machine.  This is the exact ros pub i used:

`rostic pub battery_temps geometry_msgs/Point -- -3000 62 90`

vashmata commented:
Okay, I pulled and tested it again and it catches negative temps/voltages after all. The actual issue is that when I get an error message for battery voltage once, any new out-of-bounds voltages don't generate an error popup. The same issue occurs with the battery temperatures but the voltages and temperatures cases are independent. Refreshing the page resets it and I can get the error again.

Also, I can hear the sound from the popup now

adrientremblay commented:
Yeah I kind of wrote it that way.  So if theres an error popup open for one temp or for the voltage then subsequent errors wont create a new one.  Maybe I'll change it so that when another error triggers it will effectively close the previous one trigger the new one.  I don't think it's a good idea to have multiple popups open though in the case that a ton of errors trigger and the user wastes time closing them all :/

adrientremblay commented:
Wait is does that already lol.  Let me know if thats sounds good.

vashmata commented:
yea everything's gucci

adrientremblay commented:
@PeterGhimself Everything works on my end which is very weird.  I ran 

```
rostopic pub /battery_temps geometry_msgs/Point 100 100 100
rostopic pub /battery_temps geometry_msgs/Point 80 80 80
rostopic pub /battery_temps geometry_msgs/Point 100 100 100

```
The fact that you were even able to get that -123 -123 -132 temp showing white means something very wrong though.  I'm kinda stumped as to what to do now though tbh since i cant replicate the issue for some reason .

Also I'll start implementing the voltage thing.  But shouldent we have ranges for when it approaches the minimum voltage too?  And have the same thing for the temperatures?

adrientremblay commented:
@MartensCedric  @PeterGhimself 
I updated the algorithm to include the warning zone.  Right now the zone is 10 degrees before the limits for tempeature and one volt before the limits for voltage.  I also added a local testing guide.

adrientremblay commented:
@PeterGhimself @MartensCedric @vashmata If there are any problems with this algo.  I'm just gonna give up and implement a timer lol.  What i should have done to begin with probably
